pkg/asset/fsck.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package asset
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+)
+
+// detectFileSystemType uses blkid to identify the file system type of a given image file
+func detectFileSystemType(imagePath string) (string, error) {
+	cmd := exec.Command("blkid", "-o", "value", "-s", "TYPE", imagePath)
+	output, err := cmd.Output()
+	if err != nil {
+		return "", fmt.Errorf("failed to detect file system type: %v", err)
+	}
+	return string(output), nil
+}
+
+// runFsck uses the fs-specific fsck subcommand to check if an image file is clean
+func runFsck(imagePath string, fsType string) (bool, error) {
+	fsckCommand := fmt.Sprintf("fsck.%s", fsType)
+	cmd := exec.Command(fsckCommand, "-n", imagePath)
+	cmd.Run()
+
+	exitCode := cmd.ProcessState.ExitCode()
+	return exitCode == 0, nil
+}
+
+// Fsck returns whether a filesystem is clean or not
+func Fsck(r io.Reader) (bool, error) {
+	// Write the image to a temporary file
+	tempFile, err := os.CreateTemp("", "*.img")
+	if err != nil {
+		return false, fmt.Errorf("failed to create temporary file: %v", err)
+	}
+	defer os.Remove(tempFile.Name())
+
+	_, err = io.Copy(tempFile, r)
+	if err != nil {
+		return false, fmt.Errorf("failed to write data to temporary file: %v", err)
+	}
+
+	if err := tempFile.Close(); err != nil {
+		return false, fmt.Errorf("failed to close temporary file: %v", err)
+	}
+
+	// Detect the fs type
+	fsType, err := detectFileSystemType(tempFile.Name())
```

Comment:
Can't  fsck auto-detect fs type?

Reply:
I checked how fsck is implemented and indeed it only supports extracting the fstype of a mounted file system https://github.com/util-linux/util-linux/blob/master/disk-utils/fsck.c#L913 so we need to implement this ourselves somehow for image files